### PR TITLE
Add config.json handling

### PR DESCRIPTION
**Pull request by ChriMarMe**
Add kexec of kernel and initramfs with args given via config.json
Signed-off-by: Christopher Meis <christopher.meis@9elements.com>

### CODE REVIEW

**Comment by ChriMarMe** [comment]
`On branch 9eBootUpdate
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   go.sum
	modified:   vendor/modules.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	vendor/github.com/xi2/

no changes added to commit (use "git add" and/or "git commit -a")
vendor/ is out-of-date: run `go mod tidy && go mod vendor` and then check in the changes
If `go mod tidy && go mod vendor` results in no changes, make sure you are using the latest relase of Go
 M vendor/modules.txt
?? vendor/github.com/xi2/`

I did `go mod tidy && go mod vendor` - I use "go version go1.14.10 linux/amd64"

I don't understand the problem.

**Comment by ChriMarMe** [comment]
Addition: 
There are no untracked files in my branch

File: cmds/boot/netdiskboot/netdiskBoot.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	cfg, err := loadConfig(cfgfile)
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+	log.Printf("\n%v", cfg.String())
+
+	log.Print(banner)
+
+	configureDHCPNetwork()
```
Comment (walterchris) [inline cmds/boot/netdiskboot/netdiskBoot.go:100]: error check?

File: cmds/boot/netdiskboot/netdiskBoot.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+	if !*doDebug {
+		log.Printf("Downloading file from %s\n", cfg.ImgURL)
+
+		err = DownloadFile("/tmp/fedora.img.xz", cfg.ImgURL)
```
Comment (walterchris) [inline cmds/boot/netdiskboot/netdiskBoot.go:105]: This should not be hardcoded - either use general name - or use name from URL.

File: cmds/boot/netdiskboot/netdiskBoot.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		log.Fatal(err)
+		return
+	}
+	log.Printf("\n%v", cfg.String())
+
+	log.Print(banner)
+
+	configureDHCPNetwork()
+
+	if !*doDebug {
```
Comment (walterchris) [inline cmds/boot/netdiskboot/netdiskBoot.go:102]: That whole part only is executed in debug mode? Maybe dry run?

File: cmds/boot/netdiskboot/netdiskBoot.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    \  /   \ \             _/ /
+    ] ]     )_\_         /__\/
+    /_\     ]___\
+   (___)
+`
+var debug = func(string, ...interface{}) {}
+
+func main() {
+	if *doDebug {
+		debug = log.Printf
```
Comment (walterchris) [inline cmds/boot/netdiskboot/netdiskBoot.go:87]: I do not see this used once here. You might want to refactor the printf's if some of these are not meant to be debug.

File: cmds/boot/netdiskboot/network.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	"fmt"
+	"log"
+	"net"
+	"time"
+
+	"github.com/u-root/u-root/pkg/dhclient"
+	"github.com/vishvananda/netlink"
+)
+
+func configureDHCPNetwork() error {
```
Comment (walterchris) [inline cmds/boot/netdiskboot/network.go:15]: I saw this function multiple times now within the code base.

File: cmds/boot/netdiskboot/network.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	for result := range r {
+		if result.Err == nil {
+			return result.Lease.Configure()
+		}
+		log.Printf("dhcp response error: %v", result.Err)
+	}
+	return errors.New("no valid DHCP configuration recieved")
+}
+
+func findNetworkInterface() (netlink.Link, error) {
```
Comment (walterchris) [inline cmds/boot/netdiskboot/network.go:44]: Yup - that one as well.

File: cmds/boot/pxeboot/pxeboot.go [hunk 7]
```diff
@@ -40,6 +40,7 @@ var (
 	noLoad      = flag.Bool("no-load", false, "get DHCP response, print chosen boot configuration, but do not download + exec it")
 	noExec      = flag.Bool("no-exec", false, "download boot configuration, but do not exec it")
 	noNetConfig = flag.Bool("no-net-config", false, "get DHCP response, but do not apply the network config it to the kernel interface")
+	skipBonded  = flag.Bool("skip-bonded", false, "Skip NICs that have already been added to a bond")
```
Comment (walterchris) [inline cmds/boot/pxeboot/pxeboot.go:43]: ? Does that belong here?

File: cmds/boot/pxeboot/pxeboot.go [hunk 8]
```diff
@@ -58,6 +59,10 @@ func NetbootImages(ifaceNames string) ([]boot.OSImage, error) {
 		return nil, err
 	}
 
+	if *skipBonded {
```
Comment (walterchris) [inline cmds/boot/pxeboot/pxeboot.go:62]: Same

File: pkg/boot/grub/grub.go [hunk 9]
```diff
@@ -19,14 +19,18 @@ import (
 	"io"
```
Comment (walterchris) [inline pkg/boot/grub/grub.go:19]: Not sure why this is in here..

File: pkg/boot/grub/grub.go [hunk 10]
```diff
@@ -298,14 +317,119 @@ func (c *parser) append(ctx context.Context, config string) error {
 		arg := kv[1]
 
 		switch directive {
+		case "search.file", "search.fs_label", "search.fs_uuid":
```
Comment (walterchris) [inline pkg/boot/grub/grub.go:320]: Okay - that looks like Ryan's grub search branch. Might be my fault. Can you rebase this on top of the u-root master?

**Review by walterchris — CHANGES_REQUESTED** [review]

File: pkg/boot/grub/grub.go (hunk 10 above)
Comment (ChriMarMe) [reply pkg/boot/grub/grub.go:320]: The last 4 comments are in here, because I rebased on master. The PR is on your branch which is not in sync with master/main branch.

File: pkg/boot/grub/grub.go (hunk 9 above)
Comment (ChriMarMe) [reply pkg/boot/grub/grub.go:19]: see last comment

File: cmds/boot/pxeboot/pxeboot.go (hunk 8 above)
Comment (ChriMarMe) [reply cmds/boot/pxeboot/pxeboot.go:62]: see last comment

File: cmds/boot/pxeboot/pxeboot.go (hunk 7 above)
Comment (ChriMarMe) [reply cmds/boot/pxeboot/pxeboot.go:43]: See last comment

File: cmds/boot/netdiskboot/network.go (hunk 5 above)
Comment (ChriMarMe) [reply cmds/boot/netdiskboot/network.go:15]: You coded it in your branch, I just moved it to a different file :P

File: cmds/boot/netdiskboot/network.go (hunk 6 above)
Comment (ChriMarMe) [reply cmds/boot/netdiskboot/network.go:44]: That one as well :D

File: cmds/boot/netdiskboot/netdiskBoot.go (hunk 3 above)
Comment (ChriMarMe) [reply cmds/boot/netdiskboot/netdiskBoot.go:102]: done

File: cmds/boot/netdiskboot/netdiskBoot.go (hunk 4 above)
Comment (ChriMarMe) [reply cmds/boot/netdiskboot/netdiskBoot.go:87]: done

File: cmds/boot/netdiskboot/netdiskBoot.go (hunk 1 above)
Comment (ChriMarMe) [reply cmds/boot/netdiskboot/netdiskBoot.go:100]: done

File: cmds/boot/netdiskboot/network.go (hunk 5 above)
Comment (walterchris) [reply cmds/boot/netdiskboot/network.go:15]: Let me rephrase that - can you clean this up ? ;)

Comment (ChriMarMe) [reply cmds/boot/netdiskboot/network.go:15]: I took a look over the different bootloaders and there is more to refactor possible in general.
I would suggest a refactoring of dhcp implementation across all bootloaders in a seperat PR to the main branch.
